### PR TITLE
fix: allow child radio buttons to be changed multiple times from javascript

### LINKED ISSUE #3903: Externally changing checked attribute in fast-components/fast-radio causes checked state to stop updating
**Describe the bug; what happened?**
When changing the checked attribute via `<fast-radio ... ?checked=${x => x.shouldBeChecked}>...</fast-radio>` between multiple options after a single round the later options will no longer accept the checked state.

**What are the steps to reproduce the issue?**
Using the provided example:
1. Click **With Selected Three**
2. Option **Three** should be selected
3. Click **With Selected One**
4. Option **One** should be selected
5. Click **With Selected Three**
6. Option **Three** is not selected, but should be

**What behavior did you expect?**
Changing the `?selected` value should cause the radio to be selected

**If applicable, provide screenshots:**
Initial State
Option Three Selected When Button Clicked
Option One Selected When Button Clicked
Option Three not Selected When Button Clicked

**In what environment did you see the issue?**

* MacOS
* Firefox 80.0.1 and Chrome 85.0.4183.102

**Is there any additional context?**
Reproduction:
[fast-radio-issue.zip](https://github.com/microsoft/fast/files/5227322/fast-radio-issue.zip)

### PR DESCRIPTION
**Pull request by marjonlynch**
# Description
If a developer tried to change a particular radio button in a radio group multiple times it would fail to change after the first change.

For input type='radio' there is a bit of a disconnect between the property checked and the attribute checked which is different than most of the other components (though Checkbox has this same behavior). 
* initially when radio group loads we need to look at the last radio with checked attribute and set the property to checked on that radio and set the checked property(NOT the attribute) to false on all the others.
* once that checked property is set, it can only be changed through the property and there is a disconnect between the checked attribute and the checked property.

Also fixed tabindexes when values are changed via code, they were not updated appropriately, only when using mouse interaction.

* still need to add specs

closes #3903 

## Issue type checklist

- [ ] **Chore**: A change that does not impact distributed packages.
- [x] **Bug fix**: A change that fixes an issue, link to the issue above.
- [ ] **New feature**: A change that adds functionality.

**Is this a breaking change?**
- [ ] This change causes current functionality to break.

**Adding or modifying component(s) in `@microsoft/fast-components` checklist**

- [ ] I have added a new component
- [x] I have modified an existing component
- [ ] I have updated the [definition file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#definition)
- [x] I have updated the [configuration file](https://github.com/Microsoft/fast/blob/master/packages/web-components/fast-components/CONTRIBUTING.md#configuration)

## Process & policy checklist

- [ ] I have added tests for my changes.
- [x] I have tested my changes.
- [ ] I have updated the project documentation to reflect my changes.
- [x] I have read the [CONTRIBUTING](https://github.com/Microsoft/fast/blob/master/CONTRIBUTING.md) documentation and followed the [standards](https://www.fast.design/docs/community/code-of-conduct/#our-standards) for this project.

### CODE REVIEW

File: packages/web-components/fast-foundation/src/radio/radio.ts [hunk 1]
```diff
@@ -101,17 +101,20 @@ export class Radio extends FormAssociated<HTMLInputElement> implements RadioCont
...
     public checked: boolean = this.defaultChecked;
     private checkedChanged(): void {
-        if (!this.dirtyChecked) {
-            this.dirtyChecked = true;
-        }
+        // TODO: temporarily removed as it's causing issues with
+        // changing the value via code and from radio-group
+        // if (!this.dirtyChecked) {
+        //     this.dirtyChecked = true;
+        // }
```
Comment (chrisdholt) [inline packages/web-components/fast-foundation/src/radio/radio.ts:108]: Is this still temporarily removed? If so, can we have something tracking it?

File: packages/web-components/fast-foundation/src/radio-group/radio-group.ts [hunk 2]
```diff
@@ -87,6 +87,16 @@ export class RadioGroup extends FASTElement {
      */
     @attr
     public value: string;
+    protected valueChanged(): void {
```
Comment (chrisdholt) [inline packages/web-components/fast-foundation/src/radio-group/radio-group.ts:90]: Can you add tests for setting the value of the group and the correct radio being selected?

Comment (marjonlynch) [reply packages/web-components/fast-foundation/src/radio-group/radio-group.ts:90]: yes, will do, plan to add a few tests for these changes

File: packages/web-components/fast-foundation/src/radio/radio.ts (hunk 1 above)
Comment (marjonlynch) [reply packages/web-components/fast-foundation/src/radio/radio.ts]: nope, added it back in now

File: packages/web-components/fast-foundation/src/radio-group/radio-group.spec.ts [hunk 3]
```diff
@@ -266,6 +266,42 @@ describe("Radio Group", () => {
         await disconnect();
     });
 
+    it("should mark the last radio defaulted to checked as checked, the rest should not be checked", async () => {
+        const { element, connect, disconnect } = await fixture(html<FASTRadioGroup>`
```
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/src/radio-group/radio-group.spec.ts:270]: You don't need the type on the html function in this case. That would only be needed if you had a special view model that you wanted to test *binding* the radio group to.

File: packages/web-components/fast-foundation/src/radio-group/radio-group.ts [hunk 4]
```diff
@@ -139,7 +149,31 @@ export class RadioGroup extends FASTElement {
             this.parentToolbar !== undefined && this.parentToolbar !== null;
     }
 
+    public disconnectedCallback(): void {
+        this.removeEventListener("keydown", this.keydownHandler);
```
Comment (EisenbergEffect) [inline packages/web-components/fast-foundation/src/radio-group/radio-group.ts:153]: If these event handlers were added directly to the template itself, you wouldn't need to have manual code to remove them.

**Review by EisenbergEffect — APPROVED** [review]
A few minor things that could be adjusted. Nothing major so good to go. Would be good to get a review from Nick since he's been looking into various approaches with form controls in general.

File: packages/web-components/fast-foundation/src/radio-group/radio-group.ts (hunk 2 above)
Comment (marjonlynch) [reply packages/web-components/fast-foundation/src/radio-group/radio-group.ts:90]: tests added, also for the multiple checked child radio situation

File: packages/web-components/fast-foundation/src/radio-group/radio-group.ts (hunk 4 above)
Comment (marjonlynch) [reply packages/web-components/fast-foundation/src/radio-group/radio-group.ts]: moved handlers to template, needed to return true on focusOutHandler but all working again now

File: packages/web-components/fast-foundation/src/radio-group/radio-group.spec.ts (hunk 3 above)
Comment (marjonlynch) [reply packages/web-components/fast-foundation/src/radio-group/radio-group.spec.ts]: I removed it for my specs, as I don't need to cast them specifically to FASTRadioGroup.

File: packages/web-components/fast-foundation/src/radio-group/radio-group.template.ts [hunk 5]
```diff
@@ -11,6 +11,10 @@ export const RadioGroupTemplate = html<RadioGroup>`
         role="radiogroup"
         aria-disabled="${x => x.disabled}"
         aria-readonly="${x => x.readOnly}"
+        @keypress="${(x, c) => x.keypressHandler(c.event as KeyboardEvent)}"
+        @click="${(x, c) => x.clickHandler(c.event as MouseEvent)}"
+        @keydown="${(x, c) => x.keydownHandler(c.event as KeyboardEvent)}"
```
Comment (scomea) [inline packages/web-components/fast-foundation/src/radio-group/radio-group.template.ts:16]: We can't just use one of keydown/keypress?  Why both?

Comment (marjonlynch) [reply packages/web-components/fast-foundation/src/radio-group/radio-group.template.ts:16]: well that seems like a valid question, looking into it

Comment (marjonlynch) [reply packages/web-components/fast-foundation/src/radio-group/radio-group.template.ts:15]: Well it was never getting called anyway as far as I can tell, so just removed it, also fixed the tabbing bug that was caused by me moving event handlers to template. That gets me every time, wondering if that's the default we really want.

**Review by scomea — APPROVED** [review]

**Comment by EisenbergEffect** [comment]
@marjonlynch Looks like this is good to go 😄

**Pull request merged**